### ISSUE TITLE
Backslash probably should be treated as an escape character

### ISSUE THREAD

**Issue by vilyus** [pacifica/python-jsonpath2#30]
### jsonpath2 version
0.4.0

### Platform Details
```sys.platform, sys.version == ('win32', '3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 21:26:53) [MSC v.1916 32 bit (Intel)]')```

### Scenario:
How to query the folowing properties: `"` (double quote) and `\` (backslash)

### Steps to Reproduce:
Though some keys containing backslash and double quote can be queried:

    >>> [m.current_value for m in Path.parse_str(r'''$["\\\""]''').match({r'\\\"': 42})]
    [42]
    >>> [m.current_value for m in Path.parse_str(r'''$["\\"]''').match({'\\\\': 42})]
    [42]

the fact that backslashes are not treated as an escape symbol probably makes some keys impossible to query.

### Expected Result:
Escaped backslash should match a single backslash in a key:

    >>> [m.current_value for m in Path.parse_str(r'''$["\\"]''').match({'\\': 42})]
    [42]

Escaped double quote should match a double quote in a key

    >>> [m.current_value for m in Path.parse_str(r'''$["\""]''').match({'"': 42})]
    [42]

### Actual Result:
Escaped backslash does not match a single backslash in a key:

    >>> [m.current_value for m in Path.parse_str(r'''$["\\"]''').match({'\\': 42})]
    [42]

but instead matches double backslashes in a key:

    >>> [m.current_value for m in Path.parse_str(r'''$["\\"]''').match({'\\\\': 42})]
    [42]

thus matching a single backslash seems to be impossible:

    >>> [m.current_value for m in Path.parse_str(r'''$["\"]''').match({'\\': 42})]
    line 1:2 token recognition error at: '"\"]'
    File "antlr4\atn\LexerATNSimulator.py", line 99, in match
      return self.execATN(input, dfa.s0)
    File "antlr4\atn\LexerATNSimulator.py", line 191, in execATN
      return self.failOrAccept(self.prevAccept, input, s.configs, t)
    File "antlr4\atn\LexerATNSimulator.py", line 250, in failOrAccept
      raise LexerNoViableAltException(self.recog, input, self.startIndex, reach)
    antlr4.error.Errors.LexerNoViableAltException: LexerNoViableAltException('"')
    
    During handling of the above exception, another exception occurred:
    
    File "antlr4\BufferedTokenStream.py", line 123, in fetch
      t = self.tokenSource.nextToken()
    File "antlr4\Lexer.py", line 130, in nextToken
      self.notifyListeners(e)		# report error
    File "antlr4\Lexer.py", line 285, in notifyListeners
      listener.syntaxError(self, None, self._tokenStartLine, self._tokenStartColumn, msg, e)
    File "antlr4\error\ErrorListener.py", line 60, in syntaxError
      delegate.syntaxError(recognizer, offendingSymbol, line, column, msg, e)
    File "jsonpath2\parser\__init__.py", line 63, in syntaxError
      raise ValueError('line {}:{} {}'.format(line, column, msg))
    ValueError: line 1:2 token recognition error at: '"\"]'

Escaped double quote matches a double quote preceded by a backslash: 

    >>> [m.current_value for m in Path.parse_str(r'''$["\""]''').match({'\\"': 42})]
    [42]

**Comment by dmlb2000** [comment]
@markborkum would this require a language (`JSONPath.g4`) change?

**Comment by markborkum** [comment]
@dmlb2000, No. The JSON portion of the JSONPath grammar is reused from the ANTLRv4 example, https://github.com/antlr/grammars-v4/blob/master/json/JSON.g4. Escapes are handled by the "ESC" fragment (see https://github.com/pacifica/python-jsonpath2/blob/master/jsonpath2/parser/JSONPath.g4#L140 for details).

@vilyus, Since the JSONPaths are static, instead of using the parser, have you tried constructing them manually? For example,
```python3
Path(RootNode(SubscriptNode(TerminalNode(), [ObjectIndexSubscript('"')])))
```
for the double-quote character and
```python3
Path(RootNode(SubscriptNode(TerminalNode(), [ObjectIndexSubscript('\\')])))
```
for the backslash character.

**Comment by vilyus** [comment]
@markborkum  manually constructed `Path`s  from your examples work and match `"` and `\` keys.

The problem is if you `str()` that path and then `Path.parse_str()` it again - you get different `Path`:

```python
>>> def escaped_quote_path():
...     return Path(RootNode(SubscriptNode(TerminalNode(), [ObjectIndexSubscript('"')])))
...
>>> str(escaped_quote_path()) == str(Path.parse_str(str(escaped_quote_path())))
False
>>> str(escaped_quote_path()), str(Path.parse_str(str(escaped_quote_path())))
('$["\\""]', '$["\\\\\\""]')
```

**Comment by dmlb2000** [comment]
@vilyus we pulled the JSON grammar portions directly from the antlr grammar, my guess is that this issue is also present with their grammar https://github.com/antlr/grammars-v4/blob/master/json/JSON.g4

I believe they need to fix the issue in their grammar first before can incorporate the changes. Can you submit your bug there?
